File: problems/2-medium/150-Evaluate-Reverse-Polish-Notation/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;


int evalRPN(vector<string>& tokens) {
    vector<int> stack;
    unordered_set<string> operators = {"+", "-", "*", "/"};

    for (string s : tokens) {
        if (operators.find(s) != operators.end()) {
            int right = stack.back();
            stack.pop_back();
            int left = stack.back();
            stack.pop_back();

            if (s == "+") {
                stack.push_back(left + right);
            } else if (s == "-") {
                stack.push_back(left - right);
            } else if (s == "*") {
                stack.push_back(left * right);
            } else {
                stack.push_back(left / right);
            }
        } else {
            stack.push_back(stoi(s));
        }
    }
    return stack.back();
}

File: problems/2-medium/150-Evaluate-Reverse-Polish-Notation/solution.py
from typing import List


def evalRPN(tokens: List[str]) -> int:
    stack = []
    my_dict = {"+", "-", "*", "/"}

    for value in tokens:
        if value not in my_dict:
            stack.append(int(value))
        else:
            val1 = stack.pop()
            val2 = stack.pop()

            if value == "+":
                stack.append(val2 + val1)
            elif value == "-":
                stack.append(val2 - val1)
            elif value == "*":
                stack.append(val2 * val1)
            else:
                stack.append(int(val2 / val1))

    return stack[0]
